src/realm/sync/network/default_socket.cpp

```diff
@@ -371,10 +378,177 @@ void DefaultWebSocketImpl::initiate_websocket_handshake()
+        m_thread.join();
+    }
+    // Shutting down - no need to lock mutex before check
+    REALM_ASSERT(m_state == State::Stopped);
+}
+
+void DefaultSocketProvider::start()
+{
+    std::unique_lock<std::mutex> lock(m_mutex);
+    // Has the thread already been started or is running
+    if (m_state == State::Starting || m_state == State::Started || m_state == State::Running)
+        return; // early return
+
+    // If the thread has been previously run, make sure it has been joined first
+    if (m_state == State::Stopping || m_state == State::Stopped) {
+        if (m_thread.joinable()) {
+            lock.unlock();
+            m_thread.join();
+            lock.lock();
+        }
+    }
+
+    m_logger_ptr->trace("Default event loop: start()");
+    REALM_ASSERT(m_state == State::Stopped);
+    do_state_update(State::Starting);
+    m_thread = std::thread{make_event_loop()};
+    lock.unlock();
+    // Wait for the thread to start before continuing
+    state_wait_for(State::Started);
+}
+
+std::function<void()> DefaultSocketProvider::make_event_loop()
+{
+    return [this]() {
+        m_logger_ptr->trace("Default event loop: thread running");
+        std::unique_lock<std::mutex> lock(m_mutex);
+        // The thread has started - don't return early until after did_create_thread()
+        if (m_state == State::Starting) {
+            do_state_update(State::Started);
+        }
+        lock.unlock();
+
+        auto will_destroy_thread = util::make_scope_exit([&]() noexcept {
+            m_logger_ptr->trace("Default event loop: thread exiting");
+            if (g_binding_callback_thread_observer)
+                g_binding_callback_thread_observer->will_destroy_thread();
+
+            {
+                std::lock_guard<std::mutex> lock(m_mutex);
+                if (m_state != State::Stopping)
+                    return; // early return
+
+                // Reset the service so it can be run again later
+                m_service.reset();
+                do_state_update(State::Stopped);
+            }
+        });
+        if (g_binding_callback_thread_observer)
+            g_binding_callback_thread_observer->did_create_thread();
+
+        lock.lock();
+        if (m_state != State::Started) {
+            // Stop has already been requested - exit early
+            return;
+        }
+        do_state_update(State::Running);
+        lock.unlock();
+
+        m_logger_ptr->trace("Default event loop: service run");
+        try {
+            m_service.run(); // Throws
+        }
+        catch (const std::exception& e) {
+            m_logger_ptr->error("Default event loop exception: ", e.what());
+            if (g_binding_callback_thread_observer)
+                g_binding_callback_thread_observer->handle_error(e);
+            else
+                throw;
+        }
+    };
+}
+
+void DefaultSocketProvider::stop(bool wait_for_stop)
+{
+    {
+        std::unique_lock<std::mutex> lock(m_mutex);
```

Comment:
since you hold this lock for the whole function, you probably don't need the extra set of braces on 493/511